### DISCUSSION TITLE
ART on different precision

### DISCUSSION THREAD

**Discussion by xplore-1** [Trusted-AI/adversarial-robustness-toolbox#2310] [Q&A] [answered]
I am trying to use pre-trained weights of CNN model on different precision like FP16 , FP64 as well as mixed precision, to test attacks and defense in PyTorch framework. 

For FP16 the error coming as : "RuntimeError: Input type (float) and bias type (c10::Half) should be the same"
For FP64 it is "cuda out of memory"
For AMP or mixed it is : "img /= 255.0
numpy.core._exceptions._UFuncOutputCastingError: Cannot cast ufunc 'divide' output from dtype('float64') to dtype('uint8') with casting rule 'same_kind'"

Does ART supports different precision model or just FP32?

**Comment by beat-buesser** [comment, marked as answer]
Hi @xplore-1 In general, ART should support multiple precisions, but it is not tested specifically. There is a variable at `art.ART_NUMPY_DTYPE` which defines the precision for internal Numpy operations. Its default is `numpy.float32`.
